### PR TITLE
improve mfs republisher

### PR DESCRIPTION
**Pull request by gammazero**
- Get updated values while retrying to publish
- Prefer Close to lifecycle context
- Do not require call to Start separate from New

### CODE REVIEW

File: mfs/repub.go [hunk 1]
```diff
@@ -123,19 +141,20 @@ func (rp *Republisher) Run(lastPublished cid.Cid) {
 				break
 			}
 
-			// If we aren't already waiting to publish something,
-			// reset the long timeout.
+			// If mot already waiting to publish something, reset the long
```
Comment (hsanjuan) [inline mfs/repub.go]: mot?

**Review by hsanjuan — APPROVED** [review]
part of me wonders if something similar cannot be achieved with a single timer that is reset for long/short intervals depending how the publishing went.

**Comment by gammazero** [comment]
> part of me wonders if something similar cannot be achieved with a single timer that is reset for long/short intervals depending how the publishing went.

As far as I can see both timers are needed. It like when batching by count: you need to know both the batch size and how many values are in the batch so far. This is batching, but with time.

The quick timer (waiting for additional results) keeps getting reset when another result arrives. If new results are coming in continuously, resetting quick timer each time, there needs to be another timer running that will trigger an update at some point even though new values keep arriving.

In other words, the quick timeout means there are no more values to put into the "batch", so do update. The long timeout means there are that the "batch" is full, so do update, even though there are still values (no quick timeout yet) arriving.
